src/test/java/ArrayIteratorTest.java

```diff
@@ -1,154 +1,94 @@
-    //      Capacity is the number of available indices in the array
-
-    @Test
-    void hasNext_capacityZeroArray_returnsFalse() {
-        Integer[] a = {};
-        Iterator<Integer> it = new ArrayIterator<>(a, 0);
-        assertFalse(it.hasNext());
-    }
-
-    @Test
-    void hasNext_fullCapacityOneArray_returnsTrue() {
-        Integer[] a = {10};
-        Iterator<Integer> it = new ArrayIterator<>(a, 1);
-        assertTrue(it.hasNext());
-    }
-
-    @Test
-    void hasNext_sizeZeroCapacityTenArray_returnsFalse() {
-        Integer[] a = {null, null, null, null, null, null, null, null, null, null};
-        Iterator<Integer> it = new ArrayIterator<>(a, 0);
-        assertFalse(it.hasNext());
-    }
-
-    @Test
-    void hasNext_sizeOneCapacityTenArray_returnsTrue() {
-        Integer[] a = {10, null, null, null, null, null, null, null, null, null};
-        Iterator<Integer> it = new ArrayIterator<>(a, 1);
-        assertTrue(it.hasNext());
-    }
-
-    @Test
-    void hasNext_sizeFiveCapacityTenArray_returnsTrue() {
-        Integer[] a = {10, 11, 12, 13, 14, null, null, null, null, null};
-        Iterator<Integer> it = new ArrayIterator<>(a, 5);
-        assertTrue(it.hasNext());
-    }
-
-    @Test
-    void hasNext_sizeOneCapacityTenArrayAfterHasNext_doesNotChangeState() {
-        Integer[] a = {10, null, null, null, null, null, null, null, null, null};
-        Iterator<Integer> it = new ArrayIterator<>(a, 1);
-        it.hasNext();
-        assertTrue(it.hasNext());
-        assertEquals(10, it.next());
-    }
-
-    @Test
-    void hasNext_sizeFiveCapacityTenArrayAfterHasNext_doesNotChangeState() {
-        Integer[] a = {10, 11, 12, 13, 14, null, null, null, null, null};
-        Iterator<Integer> it = new ArrayIterator<>(a, 5);
-        it.hasNext();
-        assertTrue(it.hasNext());
-        assertEquals(10, it.next());
-    }
-
-    @Test
-    void hasNext_endOfSizeOneCapacityTenArray_returnsFalse() {
-        Integer[] a = {10, null, null, null, null, null, null, null, null, null};
-        Iterator<Integer> it = new ArrayIterator<>(a, 1);
-        it.next();
-        assertFalse(it.hasNext());
-    }
-
-    @Test
-    void hasNext_endOfSizeFiveCapacityTenArray_returnsFalse() {
-        Integer[] a = {10, 11, 12, 13, 14, null, null, null, null, null};
-        Iterator<Integer> it = new ArrayIterator<>(a, 5);
-        for (int i = 0; i < 5; ++i) {
-            it.next();
-        }
-        assertFalse(it.hasNext());
-    }
-
-    @Test
-    void hasNext_endOfFullCapacityTenArray_returnsFalse() {
-        Integer[] a = {10, 11, 12, 13, 14, 15, 16, 17, 18, 19};
-        Iterator<Integer> it = new ArrayIterator<>(a, 10);
-        for (int i = 0; i < 10; ++i) {
-            it.next();
+    @Nested
+    class WhenEmpty {
+        @Test
+        void hasNext_empty_returnsFalse() {
+            ArrayIterator<Integer> iterator = new ArrayIterator<>(new Integer[]{null, null, null, null, null}, 0);
```

Comment:
This isn't empty really, it has size zero. There's two types of empty here, one with size zero and one with a length 0 array. (assuming the array itself is non-null).

This size mismatch can also occur with any array cardinality. There's actually 6 kinds of states to test here, each of them contains many tests
* size = 0, array.length = ???
* size = ???, array.length = 0
* size = 1, array.length = ???
* size = ???, array.length = 1
* size > 1, array.length = ???
* size = ???, array.length > 1

With each ??? being "any of the other possible values." `size = 0, array.length = ???` for instance is 3 tests, `(0,0), (0,1), (0, >1)`. Many of these tests will overlap, but those are the kinds of states to cover.

Reply:
Some of these may be invalid states, for example: 

```java
        @Test
        void hasNext_singletonCapacityZero_returnsTrue() {
            ArrayIterator<Integer> iterator = new ArrayIterator<>(new Integer[]{}, 1);
            assertTrue(iterator.hasNext());
        }

        @Test
        void next_singletonCapacityZero_throwsNoSuchElementException() {
            ArrayIterator<Integer> iterator = new ArrayIterator<>(new Integer[]{}, 1);
            assertThrows(NoSuchElementException.class, () -> iterator.next());  // <-- Actually throws index out of bounds
        }

```

Reply:
> * The invalid states are not included (empty has 0, 1, many size arrays, single has 1 and many, many has many)

Righto.

Tests are the first thing that actually use the code being tested. As such, their purpose is (at least) two-fold: proving correctness and depicting example usage. If you find the code being tested is hard to test, it's probably because it's hard to use and understand. If you find yourself questioning the validity of the tests, it's more likely the source is the problem and not the tests. If the class in question can be in an invalid state, maybe consider validating it beforehand; something like that. Note that this is **not** an argument for removing the tests, rather changing what piece of a class is tested for that invalid state.

This situation is a little odd, because you're testing a _private_ class. It's not literally private, but one would never use the `ArrayIterator` directly, they'd use the collection type. So, you can argue that this class being a little hard to use is ok. As mentioned above though, you absolutely should depict the behaviour of the methods under test in all situations. Make it obvious that the class can be created in an invalid state, and what happens in that invalid state.

Reply:
To follow up, given:

```java
        @Test
        void hasNext_singletonCapacityZero_returnsTrue() {
            ArrayIterator<Integer> iterator = new ArrayIterator<>(new Integer[]{}, 1);
            assertTrue(iterator.hasNext());
        }

        @Test
        void next_singletonCapacityZero_throwsNoSuchElementException() {
            ArrayIterator<Integer> iterator = new ArrayIterator<>(new Integer[]{}, 1);
            assertThrows(NoSuchElementException.class, () -> iterator.next());  // <-- Actually throws index out of bounds
        }
```        
`hasNext_singletonCapacityZero_returnsTrue` is fine, but `next_singletonCapacityZero_throwsNoSuchElementException` can still exist as `next_singletonCapacityZero_throwsIndexOutOfBoundsException`?

Reply:
_Or_, update the implementation:

```java
    public ArrayIterator(T[] elements, int size) {
        if (size >= elements.length) {
            throw new IllegalArgumentException("size exceeds length of array.");
        }
        this.elements = elements;
        this.size = size;
        this.index = 0;
    }
```